### PR TITLE
sam0: Implement Linux style device tree clock control

### PR DESCRIPTION
**Pull request by Sizurka**
This series moves the clocking for SAMD2x to be entirely controlled by the device tree.  This is similar to the ongoing work being done for #15363 but here the process is more invasive (it requires an clock provider definition) and the result is that the clock is controlled by the device tree too.

The primary motivation for the change was that the SAM0 peripherals tend to have relatively limited built in clock prescalers, but the SAM0 chip itself has a flexible clock generation system.  So it makes sense to make the chip clocking available in a more generic way.  For example, the SERCOM in I2C mode only has an 8-bit clock divider, so if using normal speed mode, this effectively limits the CPU frequency if it's clocked from the undivided core clock (e.g. SAMD5x).  Similarly, the ADC has only a few powers of two divisors available, so achieving near its rated conversion rate means using an alternate clock than the CPU clock.  In a case where at the least description is needed is the RTC (system timer) where SAMD2x can clock it at the full speed, SAMD5x is limited to a 32kHz clock.

## Device Tree Update

The first change made is to describe the existing clock structure with the device tree.  The clock definitions use the [Linux bindings](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/plain/Documentation/devicetree/bindings/clock/clock-bindings.txt) for clock control.  Basically these define a `clocks` parent node with various nodes for each individual clock provider/oscillator.  See [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts/lpc32xx.dtsi#n34) for an example.  

For SAM0, there isn't a direct central clock controller like in the above example, instead each individual GCLK is defined separately and the peripherals reference the originating GCLK directly.  This approach was chosen because each GCLK can have its own separate frequency, so specifying them all would get quite verbose.

So a sample of the SAM0 DTS ends up looking like:
```DTS
clocks {
	dfll: dfll {
		compatible = "fixed-clock";
		clock-frequency = <48000000>;
		clocks = <&gclk1>;
		#clock-cells = <0>;
		label = "DFLL";
	};

	gclk0: gclk0 {
		compatible = "fixed-clock";
		clock-frequency = <48000000>;
		clocks = <&dfll>;
		#clock-cells = <0>;
		label = "GCLK_0";
	};
};
soc {
	sercom0: sercom@42000800 {
		compatible = "atmel,sam0-sercom";
		reg = <0x42000800 0x40>;
		status = "disabled";
		label = "SERCOM0";
		clocks = <&gclk0>;
	};
};
```

The above describes two clocks, the DFLL that provides a 48MHz "source" clock, with its own (not shown) reference clock input and GCLK0, which is what peripherals can actually reference directly.  The SERCOM peripheral then uses GCLK0 as the clock input to the device.

This series incrementally converts SAM0 to use this kind of device tree.  Initially the clock setup is only described with the device tree.  Then peripherals are converted to use information from it, instead of simply assuming a constant setup.  Finally, the SoC setup code is changed to actually allow for re-configuration of the clock tree based on the device tree.

## Implementation

The first part of the implementation is simply describing the clock tree as it is used in the various boards.  Since there are Kconfig options that change this somewhat, the initial setup assumes the board defaults.  Thankfully, these Kconfig options have no peripheral visible effects, so there are no inconsistent behaviors.

Then a `clock_control` driver is implemented that can be used to query and manipulate peripheral GCLK connections.  This can be used to replace the current hard coded GCLK enables that all peripherals use.  It also has a frequency query that will be used to replace the current static defines that the peripherals use.

Next, peripherals are converted to use the clock control driver.  Since the current clock setup simply describes the existing GCLK tree (i.e. nearly everything hangs off GCLK0), this has no current behavior change.

Finally, the SoC setup is re-written to use the DTS generated information to actually generate the clock tree.  The resulting clock tree should be identical to the currently used one, since the DTS is set to describe it already.  So the final result of the series is no functional changes, but much greater flexibility.

## Dynamic Control

After this series, it becomes possible to control the clocks from DTS overlays to allow for more flexibility.  As a toy example, the below changes the main SoC clock to use a 48MHz clock generated by the FDPLL from the calibrated 32kHz RC oscillator:
```DTS
&osc32k {
	status = "ok";
};

&gclk1 {
	clocks = <&osc32k>;
	clock-frequency = <32768>;
};

&fdpll {
	status = "ok";
	clock-frequency = <48000000>;
	clocks = <&gclk1>;
};

&gclk0 {
	clocks = <&fdpll>;
};
```

This example first turns on the calibrated oscillator, then changes GCLK1 to use it as the input, and finally enables the FDPLL (the multiplier is calculated automatically) and tells GCLK0 to use it.

## Future Work

During work on this series, one thing I noticed was that the manual specifies the maximum input clock for the DFLL in closed loop mode as 33 kHz (section 37.6, table 37-7, page 985).  However the current default clock setup for any SAMDx board without a 32kHz crystal is to use the undivided 8MHz RC oscillator as the reference.  This does appear to "work", but I'm not sure it's actually using the reference (i.e. it's probably operating in (un-calibrated) open loop mode with the "reference stopped" logic kicking in).  This appears to have been carried over/cargo culted from various online forums, where, as far as I can tell, the initial recommendation of adding a divisor to GCLK1 before using it as the reference got lost at one point.  So in the future the clock setup should either be changed to use a divided OSC8M, to simply use the internal 32kHz RC oscillator, or to simply operate in open loop mode explicitly (like it's probably doing already).

Another candidate for some clock related work is enabling USB clock recovery for the DFLL.  This is Microchip's recommendation if using the chip for USB with no crystal present, but the current setups do not do that on the boards without crystals.  That is, they simply operate in closed loop mode from the internal oscillator, which may not even work (above).  So in the future, an option could be implemented to allow for operating in this mode.

These issues are not part of this series, since the goal of it was no functional changes.

## Alternative Implementations

One possible alternative implementation uses `clock_control` drivers for all possible clocks on the SoC and relies on the init priority to handle sequencing.  This has the advantage of abstracting the clocks entirely, but ends up being rather hard to follow.  Additionally, most of the clocks (i.e. anything other than a GCLK) is basically unused after SoC startup, since the peripherals do not interact with them directly.  I have an example of this kind of implementation at Sizurka/zephyr#3.

Another possibility is to enhance the DTS parsing to extract information about the parent clocks directly.  This can allow the needed information to configure the clocks at compile time, but could be limited in some cases (e.g. if a device needed to know about the source of its source clock it becomes awkward quickly).  For strictly SAM0 devices, I tend to favor this approach, but it's also more implementation specific, so the above is my attempt at a best of both worlds approach (e.g. clock setup is mostly compile time calculated but peripherals use the device abstraction).  My original implementation of this is at Sizurka/zephyr#4.

One notable disadvantage of the device based approaches is that the clock frequency is unavailable at compile time.  So any checks/asserts on it have to be runtime where they may be missed.  Additionally this complicates counter drivers, which require their frequency to be available at compile time.  So for those, it will be necessary to assume that the requested frequency is correct, then calculate the actual frequency at run time and warn if the two don't match.

### CODE REVIEW

**Comment by pizi-nordic** [comment]
Thank you for this PR! We would like to introduce clock nodes in the same way as you presented in order to trace dependencies between peripherals and clock sources. However such approach will need a strong cooperation of the parties maintaining different SoCs - we will have to update DTS files, DTS scripts and probably drivers with their infrastructure. Could we raise this on the dev-meeting?

**Comment by benpicco** [comment]
Again, nice work!
This squashed one remaining wart with sam0 and just as I was stumbling upon it, I figured you already did propose a solution for it.

File: drivers/interrupt_controller/sam0_eic.c
Comment (tbursztyka) [inline drivers/interrupt_controller/sam0_eic.c]: always put variable declaration at the beginning of a code block.
(So here, right before ARG_UNUSED)

File: drivers/watchdog/wdt_sam0.c
Comment (tbursztyka) [inline drivers/watchdog/wdt_sam0.c]: beware of the 80 chars limit

File: drivers/interrupt_controller/sam0_eic.c
Comment (Sizurka) [reply drivers/interrupt_controller/sam0_eic.c]: Changed here and a few other places.

File: drivers/watchdog/wdt_sam0.c
Comment (Sizurka) [reply drivers/watchdog/wdt_sam0.c]: Fixed the cases where I could (by splitting the clock control device into a separate variable and/or moving the initialization out of the definition).  There's a few left that I can't without also having misleading indenting (mostly due to long HAL names combined with existing indents).

File: drivers/clock_control/clock_control_sam0_gclk.c
Comment (benpicco) [inline drivers/clock_control/clock_control_sam0_gclk.c]: Do we need to wait here until the `SYNCBUSY` bit/register is cleared?

Comment (Sizurka) [reply drivers/clock_control/clock_control_sam0_gclk.c]: As I read it: no.  The `CLKCTRL` or `PCHCTRL` registers are not listed as write-synchronized in the manuals so access shouldn't require anything special.  The various other parts of the GCLK control are, but it doesn't look like the actual gating requires synchronization.

File: dts/arm/atmel/samd21.dtsi
Comment (benpicco) [inline dts/arm/atmel/samd21.dtsi]: samd5x has two of those - would it make sense to call it `fdpll0` to make sharing code easier?
I'm currently writing a separate `soc_samd5x.c` but I want to see if if I can merge it with `soc_samd2x.c` later.

File: dts/arm/atmel/samd.dtsi
Comment (benpicco) [inline dts/arm/atmel/samd.dtsi]: samd5x also has two of these, calling it `xosc0` should make sharing code easier.

File: dts/arm/atmel/samd21.dtsi
Comment (Sizurka) [reply dts/arm/atmel/samd21.dtsi]: I am slightly inclined to keep them separate so you can do constructs like:

```C
#ifdef DT_FIXED_CLOCK_FDPLL_CLOCK_FREQUENCY
/* SAMD2x ASF referencing code */
#endif
#ifdef DT_FIXED_CLOCK_FDPLL0_CLOCK_FREQUENCY
/* SAMD5x ASF referencing code */
#endif
#ifdef DT_FIXED_CLOCK_FDPLL1_CLOCK_FREQUENCY
/* SAMD5x ASF referencing code */
#endif
```

The counter argument is that the SAMD5x ASF headers do define some aliases (e.g. `GCLK_SOURCE_FDPLL` for `GCLK_SOURCE_FDPLL0`) so some of that is unneeded (bad example, since you'd still need that one; the naming is slightly different).  

Basically in nearly all cases where this is referenced, there's going to be ASF defines, and I haven't really tried to port the SAMD5x code to figure out how much of those can be shared without an ifdef soup.

So, if you expect that the sharing would be simpler with the ifdefs in the inner blocks, I can certainly change it.

Comment (benpicco) [reply dts/arm/atmel/samd21.dtsi]: I have a [very much work in progress](https://github.com/benpicco/zephyr/tree/same54-dts-clock) port to SAM D5x.
The structure is mostly the same, but register names change all over the place, so it's probably cleaner to keep them separate for now.

File: drivers/adc/adc_sam0.c
Comment (pizi-nordic) [inline drivers/adc/adc_sam0.c]: IMHO information about clock driving given peripheral should be managed by OS, not the driver.
Ideally, we should have something like device_clock_get(dev, idx).

File: soc/arm/atmel_sam0/common/soc_samd2x.c
Comment (benpicco) [inline soc/arm/atmel_sam0/common/soc_samd2x.c]: I've only tested this on same54 so far, but there I need to source `FDPLL032K` if `FDPLL0` is sourced by a GCLK. instead of e.g. XOSC32K

Comment (benpicco) [reply soc/arm/atmel_sam0/common/soc_samd2x.c]: I'm not sure what I wrote there anymore - the effect I'm observing is that if GCLK1 is sourced by ULPOSC32K and FDPLL0 is sourced by GCLK1, it takes a couple of seconds for the FDPLL to LOCK & get READY.

Non-deterministic behavior lends itself to ill conclusions…

I think I clocked the system before from OSCULP32K without that a long startup delay, so I'll have to investigate a bit.

Comment (Sizurka) [reply soc/arm/atmel_sam0/common/soc_samd2x.c]: Do you still get the same behavior if you do something XOSC32K->GCLK1->FDPLL0?  I'm wondering if it's just instability in the ULPOSC32K causing a slow lock.

If you do end up needing it, then you can do something like:
```dts
&fdpll {
	clocks = <&gclk1>, <&gclk2>;
};
```

Then use `DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER_0` and `DT_FIXED_CLOCK_FDPLL_CLOCK_CONTROLLER_1` for the reference and `FDPLL032K`. 

It works without one on SAMD21 and the manual says "The FDPLL96M requires a 32kHz clock from the GCLK when the FDPLL96M internal lock timer is used."  So I guess it doesn't explicitly say you can operate without one, but I certainly read it like that.  The SAMD5x manual says "When the optional lock timer timeout function is used, a 32KHz reference clock (GCLK_DPLL_32K) is also required." (28.5.3, p 773), so it'd be somewhat surprising if it needed it regardless.

File: drivers/adc/adc_sam0.c
Comment (Sizurka) [reply drivers/adc/adc_sam0.c]: I agree that having that kind of API would be very useful, I'm not sure how you'd actually implement it.  You'd somehow need to attach a dynamic number of references to the `struct device` while also avoiding a combinatorial explosion with the device definitions (e.g. you already have to deal with the effects of `CONFIG_DEVICE_POWER_MANAGEMENT` on device declaration, even adding a single clock would take the permutations to four).

On a higher level, it seems like a more general "get associated device" API would be useful.  There's at least two cases of "sibling" devices currently handled like what I'm doing here (other `clock_control` usage and DMA).  I could also see an argument for having the same API get the parent and children devices on a bus.  I have no idea how you'd implement that, though.

In either case, this (and the other cases that I copied the approach from) could be easily adapted to that kind of API if it became available.

**Review by benpicco — APPROVED** [review]

**Comment by benpicco** [comment]
recheck

**Comment by benpicco** [comment]
needs rebasing again…

File: dts/arm/atmel/samd20.dtsi
Comment (galak) [inline dts/arm/atmel/samd20.dtsi]: I wonder if this should be pulled out into its own PR.  Is it the case that some SoC's don't have TC6 at all?  If so, the node shouldn't exist on those SoCs (rather than using 'disabled').

File: dts/bindings/clock/fixed-clock.yaml
Comment (galak) [inline dts/bindings/clock/fixed-clock.yaml]: I don't think we should have the inherits from base.yaml here.

**Review by galak — CHANGES_REQUESTED** [review]
A few initial comments.  In general this looks good.

File: dts/arm/atmel/samd20.dtsi
Comment (Sizurka) [reply dts/arm/atmel/samd20.dtsi]: The issue is we don't (currently) have granularity below "SAM D20", while the configurations of the chip vary between "SAM D20 J/G/E".  When it's present on a SAM D20 chip it shared the same properties (i.e. the hardware definition is common between all SAM D20, but various configurations may have it or not).  So in that way it makes sense to put it in SAM D20 top level, at least.  

SERCOM has the same issue, but it doesn't come up because they're already "disabled" by the compatible flagging.  That is, they require the overlay to set the compatible property before any defines are generated.

It's actually a bit worse than that if you get into SAM D21, since they have further configuration splitting (e.g. SAMD21E16C vs SAMD21E16L vs SAMD21G16L all have different availability between TCs and SERCOM).  So I mostly just didn't want to get into that.

So if the SAMD2x DTSI files are ever refactored, that would likely be something to incorporate.  Here I just did this because it was the simplest way to make sure the unavailable HAL constants weren't being referenced by default.  With that said, I'm not against removing the node entirely until such a refactor is done, either.

File: dts/bindings/clock/fixed-clock.yaml
Comment (Sizurka) [reply dts/bindings/clock/fixed-clock.yaml]: Ok, I've removed the inherit.

**Comment by Sizurka** [comment]
@galak WRT the DFLL from the other thread: it's a fixed frequency oscillator (48 MHz) that can optionally use a reference clock to set the trim/calibration; it just adjusts the trim register up/down based on if it gets less/more than the expected number of cycles per reference clock tick.  

However, this clearly highlights that the concept of a fixed clock with a clock input isn't sufficiently universal, so I've split it out into three vendor specific ones (`atmel,sam0-dfll`, etc).  The other option would be a different type of generic clock for ones that need a reference; probably something like Linux's `fixed-factor-clock` (maybe `fixed-reference-clock`?).

I've also changed it to use `clock-output-names` as the identifier instead of `label`, though it's only valid for cases where there's only a single output.

**Review by pizi-nordic — APPROVED** [review]

**Comment by galak** [comment]
Let's get PR #17127 which should hopefully unblock this PR.

**Comment by carlescufi** [comment]
@Sizurka could you rebase on top of master now that #17127 is merged?

**Comment by carlescufi** [comment]
@galak can you take another look please?

File: dts/bindings/clock/clock.yaml
Comment (galak) [inline dts/bindings/clock/clock.yaml]: `clock-frequency` should stay in fixed-clock.yaml.

File: dts/bindings/clock/atmel,sam0-gclk.yaml
Comment (galak) [inline dts/bindings/clock/atmel,sam0-gclk.yaml]: What is `clock-frequency` suppose to mean?

File: dts/bindings/clock/atmel,sam0-fdpll.yaml
Comment (galak) [inline dts/bindings/clock/atmel,sam0-fdpll.yaml]: What is `clock-frequency` suppose to mean?

File: dts/bindings/clock/atmel,sam0-dfll.yaml
Comment (galak) [inline dts/bindings/clock/atmel,sam0-dfll.yaml]: What is `clock-frequency` suppose to mean?

File: dts/bindings/timer/atmel,sam0-tc32.yaml
Comment (galak) [inline dts/bindings/timer/atmel,sam0-tc32.yaml]: Why isn't the clock-freq figured out from the clock for this device?

File: dts/bindings/iio/adc/atmel,sam0-adc.yaml
Comment (galak) [inline dts/bindings/iio/adc/atmel,sam0-adc.yaml]: Why isn't the clock-freq figured out from the clock for this device?

**Review by galak — CHANGES_REQUESTED** [review]
Some questions and comments on the dts bindings.

File: dts/bindings/clock/clock.yaml
Comment (Sizurka) [reply dts/bindings/clock/clock.yaml]: Ok.

File: dts/bindings/clock/atmel,sam0-gclk.yaml
Comment (Sizurka) [reply dts/bindings/clock/atmel,sam0-gclk.yaml]: This is the output clock frequency of the GCLK.  This is required because it is used to calculate the divisor the GCLK applies to its input and because knowing the input frequency would require a recursive walk otherwise (i.e. when a GCLK sources from another GCLK).

I have clarified the description on the binding.

File: dts/bindings/clock/atmel,sam0-fdpll.yaml
Comment (Sizurka) [reply dts/bindings/clock/atmel,sam0-fdpll.yaml]: This is the output frequency of the clock provider.  It is used together with the input clock frequency and the electrical limits of the device to calculate the divisor(s) and the multiplier required to produce the frequency specified. 

I have clarified the description on the binding.

File: dts/bindings/clock/atmel,sam0-dfll.yaml
Comment (Sizurka) [reply dts/bindings/clock/atmel,sam0-dfll.yaml]: This is the output frequency of the oscillator.  It is determined by the hardware (e.g. 48 MHz on SAMD2x or 32 MHz on SAML1x).

I have clarified the description on the binding.

File: dts/bindings/timer/atmel,sam0-tc32.yaml
Comment (Sizurka) [reply dts/bindings/timer/atmel,sam0-tc32.yaml]: Two reasons:

1. The TC peripheral itself has an on-peripheral divisor available, so this is used to calculate what that divisor should be (similar to how the SPI drivers calculate their clock rate).
2. The `counter` API requires that the frequency be available at compile time (`.info`), so we can't use the `clock_control` API to query it (since that's on available at run time).

I have clarified the description on the binding.

File: dts/bindings/iio/adc/atmel,sam0-adc.yaml
Comment (Sizurka) [reply dts/bindings/iio/adc/atmel,sam0-adc.yaml]: Similar to the `counter`, the ADC has an on-peripheral divisor available, which this is used to calculate.  By default, this is set to the maximum frequency the ADC can support (determined by hardware, e.g. 2.1 MHz on SAMD2x or 16 MHz on SAME54), so this is used simply to calculate the divisor needed to keep the ADC within spec. 
 But it can also be used to apply the divisor to achieve different conversion rates (e.g. setting it to 1 MHz to allow for longer sample acquisition times).

I have clarified the description on the binding.

**Comment by galak** [comment]
Was hoping to find you on slack to discuss this further.  Trying to see what values you are trying to compute based on the DTS info.  Wondering if we should be using 'assigned-clock-rates` (https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/clock/clock-bindings.txt).

**Comment by Sizurka** [comment]
I don't think `assigned-clock-rates` is a good fit here.  The Linux binding description of that is for a node that configures a parent clock and is the only consumer of that clock.  Here the clocks are either internal to the node or exist as one to many clocks so they can have multiple consumers.

Basically, my usage for `clock-frequency` in nodes was to calculate divisors and/or multipliers required for that node to produce the requested clock rate.  For the clock nodes (/clocks/* ) it is also the output frequency of the clock.  For device nodes (/soc/* and /cpus/cpu0) it is used with the input clock frequency to calculate the divisor required to produce the clock rate.

The ADC and TC32 (`counter`) have two asynchronous clocks involved: the input/register/device clock (the `clocks` property) and a device internal clock derived from the input clock.  The internal clock frequency is set with the nodes `clock-frequency` and the driver calculates the divisor from the input clock required to achieve that frequency.  The internal clock cannot be directly used outside of the device (it cannot act as a clock provider), but I'm not aware of a better "common" property name to reflect this.

The clock nodes all use `clock-frequency` as expected: it's the output frequency of the clock.  The SoC init code uses this frequency and the input frequency (if any) to calculate the required settings for the clock hardware.


For example, the branch of the clock tree driving the ADC might look like: XOSC32K -> GCLK1 -> DFLL -> GCLK4 -> ADC0

DT node         | Clock Frequency                        | Zephyr Subsystem | Calculations                         | Purpose
---             | ---                                    | ---              | ---                                  | ---
| /clocks/xosc32k | 32768 (fixed)                          | SoC init         |                                      | Hardware clock source (one to many GCLK and other clocks)
| /clocks/gclk1   | 32768 (selectable <= input)            | `clock_control`  | Divisor = 1 (32768/32768)            | Clock distribution (one to many device)
| /clocks/dfll    | 48000000 (fixed, with reference)       | SoC Init         | Multiplier = 1464 (48000000 / 32768) | Hardware clock source (one to many GCLK)
| /clocks/gclk4   | 16000000 (selectable <= input)         | `clock_control`  | Divisor = 3 (48000000 / 16000000)    | Clock distribution (one to many device)
| /soc/adc        | 16000000 - async (fixed = input)       | `adc`            |                                      | Device clock (registers, internal)
| /soc/adc        | 2000000 - sample (selectable <= input) | `adc`            | Divisor = 8 (16000000 / 2000000)     | Sampling clock (acquisition time)

Here there is an external 32KiHz oscillator (/clocks/xosc32k) which is connected to GCLK1 (/clocks/gclk1).  GCLK1 is then used as the reference frequency for the DFLL (/clocks/dfll), which it uses to adjust its calibration dynamically.  Then the DFLL is connected to GCLK4 (/clocks/gclk4), but with GCLK4 providing a divisor (/3) so that the frequency exposed on GCLK4 is 16MHz.  Finally the ADC (/soc/adc) gets its "asynchronous" clock from GCLK4 (the `clocks` property), this clock is used for register access and internal device clocking.
The ADC then has a second clock used to control the sampling hardware (i.e. the rate at which the SAR hardware is clocked).  This clock rate is derived from the asynchronous clock with a divisor calculated from the ADC's `clock-frequency` property.  So here the `clock-frequency` property controls an entirely internal clock to the device/node.

Overall, I'm not against using a different property for the device internal clock rates, but the upstream description of `assigned-clock-rates` doesn't fit.  Most (all?) of the examples of similar upstream usage I found used vendor specific properties for this (e.g. `st,adc-freq`, `fsl,adck-max-frequency`, etc).  

Unfortunately, it doesn't look like the current DTS parser can handle the comma in property names (I haven't tracked down why), so even if changed, it would still end be being somewhat different.  So I can certainly change it for the device nodes, but I do think `clock-frequency` fits the clocks nodes in this case.

Basically, would changing the ADC and counter frequency properties to (e.g.) `atmel-sam0-adc-frequency` and `atmel-sam0-tc32-frequency` be sufficient?

**Comment by galak** [comment]
Let me chat with the Linux clock maintainers to see how we should be representing things.  I have a feeling the wording around `assigned-clock-rates` is a bit confusing.

**Comment by galak** [comment]
In talking to the Linux clock maintainers I think we should have the DTS look something like:

(Note, just show GCLK and DFLL in this example)

```
                dfll: dfll {
                        compatible = "atmel,sam0-dfll";
                        assigned-clock-rates = <48000000>;
                        assigned-clocks = <&gclk 1>;
                        clocks = <&gclk 0>, <&gclk 1>, <&gclk 2>, <&gclk 3>,
                                 <&gclk 4>; <&gclk 5>, <&gclk 6>, <&gclk 7>;
                        #clock-cells = <0>;
                        clock-output-names = "DFLL";
                };

                gclk: gclk {
                        compatible = "atmel,sam0-gclk";
                        reg = <0x40000C00 0xc>;
                        clocks = <&osc32k &ulposc32k &osc8m &xosc32k &xosc>;
                        #clock-cells = <1>;
                        clock-output-names = "GCLK_0", "GCLK_1", "GCLK_2",
                                             "GCLK_3", "GCLK_4", "GCLK_5",
                                             "GCLK_6", "GCLK_7", "GCLK_8";
                };

```

**Comment by ioannisg** [comment]
@Sizurka is this a stale PR?

**Comment by Sizurka** [comment]
Yes, I was unable to come up with a way to actually generate the clock control from the requested DT structure; I'll close the PR.

**Pull request closed without merging**
